### PR TITLE
feat: [IDP-908]: Admin Scorecard APIs

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51385)

### CODE REVIEW

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -803,6 +803,162 @@ paths:
...
+        - $ref: '#/components/parameters/AccountHeader'
+        - $ref: '#/components/parameters/CheckId'
+      x-internal: true
+    put:
+      summary: Update check details for given checkId
+      tags:
+        - Checks
+      operationId: update-check
+      responses:
+        '201':
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:871]: this should not be 201 for update call

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -803,6 +803,162 @@ paths:
...
+        - $ref: '#/components/parameters/ScorecardId'
+      x-stoplight:
+        id: fr7i2p91vje24
+    put:
+      summary: Update scorecard details for given scorecardId
+      tags:
+        - Scorecards
+      operationId: update-scorecard
+      responses:
+        '201':
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:946]: same here

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 3]
```diff
@@ -1872,6 +2028,319 @@ components:
...
+          type: string
+        description:
+          type: string
+        filters:
+          x-stoplight:
+            id: f1au4kix9z5hv
+          type: array
+          items:
+            $ref: '#/components/schemas/ScorecardFilters'
+        weightage_strateg:
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:2246]: missing a "y" i believe

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 2 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 3 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: modified

File: idp-service/src/main/java/io/harness/idp/scorecard/checks/entity/CheckEntity.java [hunk 4]
```diff
@@ -51,7 +53,7 @@ public static List<MongoIndex> mongoIndexes() {
   private String name;
   private String description;
 
-  private RuleStrategy ruleStrategy;
+  private CheckDetails.RuleStrategyEnum ruleStrategy;
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/scorecard/checks/entity/CheckEntity.java:56]: Earlier I guess we have duplicated the enums and inner classes at the entity layer as well. Should we continue that or use from spec directly like this one?

Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/scorecard/checks/entity/CheckEntity.java:56]: I think its better we make use of spec directly as it removes duplicate effort to maintain enums. I have also incorporated the same in StatusInfoEntity earlier

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 5]
```diff
@@ -1872,6 +2028,319 @@ components:
       properties:
         resolved_env_variables:
           type: string
+    Check:
+      title: Check
+      x-stoplight:
+        id: ep5latvpj75jo
+      type: object
+      description: Check
+      properties:
```
Comment (vikyathharekal) [inline idp-service/contracts/openapi/v1/openapi.yaml:2037]: Can we add labels and data source as well?

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 6]
```diff
@@ -1872,6 +2028,319 @@ components:
...
+          $ref: '#/components/schemas/CheckDetails'
+      required:
+        - check_details
+    Rule:
+      title: Rule
+      x-stoplight:
+        id: qg6l73nz6jl5o
+      type: object
+      properties:
+        data_source_id:
```
Comment (vikyathharekal) [inline idp-service/contracts/openapi/v1/openapi.yaml:2140]:
```suggestion
        data_source_identifier:
```

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 7]
```diff
@@ -1872,6 +2028,319 @@ components:
...
+          type: string
+          x-stoplight:
+            id: rhssuuh5er4s3
+        data_point_id:
```
Comment (vikyathharekal) [inline idp-service/contracts/openapi/v1/openapi.yaml:2144]:
```suggestion
        data_point_identifier:
```

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 8]
```diff
@@ -1872,6 +2028,319 @@ components:
...
+          properties:
+            name:
+              type: string
+              x-stoplight:
+                id: vwp9icmaj66xs
+            description:
+              type: string
+              x-stoplight:
+                id: nkijttl2v7o11
+            expression:
```
Comment (vikyathharekal) [inline idp-service/contracts/openapi/v1/openapi.yaml:2196]: Do we need expression here? I guess we aren't showing this in the UI in the scorecard page

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 9]
```diff
@@ -1991,6 +2460,27 @@ components:
...
         type: string
       description: Namespace from where the request is being made
+    CheckType:
+      name: custom
+      in: query
+      schema:
+        type: boolean
+      description: Type of checks to be returned
+      required: true
+    CheckId:
```
Comment (vikyathharekal) [inline idp-service/contracts/openapi/v1/openapi.yaml:2470]:
```suggestion
    CheckIdentifier:
```

File: idp-service/src/main/java/io/harness/idp/scorecard/checks/entity/CheckEntity.java (hunk 4 above)
Comment (vikyathharekal) [reply idp-service/src/main/java/io/harness/idp/scorecard/checks/entity/CheckEntity.java:56]: Yes makes sense. We should also do the same for duplication we have added earlier. I'll create a ticket to track it, anyone who has some bandwidth can pick it up later.

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 10]
```diff
@@ -1872,6 +2028,319 @@ components:
...
+          x-stoplight:
+            id: zzxd0l8eme1v9
+        description:
+          type: string
+          x-stoplight:
+            id: q2uqq8ic90w76
+        checks:
+          type: array
+          items:
+            $ref: '#/components/schemas/Check'
```
[the remark this answers is not in the document]
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml:2248]: In the scorecard list view, we are displaying Check's expression. Hence we are using this check object 
```
Check:
 - identifier
 - name
 - description
 - expression
 - lables
```

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 9 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 11]
```diff
@@ -1991,6 +2460,27 @@ components:
...
+      description: Type of checks to be returned
+      required: true
+    CheckId:
+      name: check-id
+      in: path
+      description: Check Identifier
+      required: true
+      schema:
+        type: string
+    ScorecardId:
```
[the remark this answers is not in the document]
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 7 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 6 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 5 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml:2037]: done

File: idp-service/src/main/java/io/harness/idp/scorecard/checks/service/CheckServiceImpl.java [hunk 12]
```diff
@@ -9,6 +9,54 @@
...
+
+  @Override
+  public CheckDetails getCheckDetails(String accountIdentifier, String identifier) {
+    CheckEntity checkEntity = checkRepository.findByAccountIdentifierAndIdentifier(accountIdentifier, identifier);
+    return CheckDetailsMapper.toDTO(checkEntity);
+  }
+
+  @Override
+  public List<CheckEntity> getChecksByAccountIdAndIdentifiers(String accountIdentifier, List<String> identifiers) {
+    List<CheckEntity> entities = checkRepository.findByAccountIdentifierAndIdentifierIn(accountIdentifier, identifiers);
```
[the remark this answers is not in the document]
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/scorecard/checks/service/CheckServiceImpl.java]: done

File: idp-service/contracts/openapi/v1/openapi.yaml (hunk 8 above)
Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: removed

**Review by vikyathharekal — DISMISSED** [review]

**Review by mankrit-singh — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
trigger review

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger ut2, ut6, ut7

**Review by mankrit-singh — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
trigger uts

**Pull request merged**